### PR TITLE
D2 vscode error experience

### PR DESCRIPTION
**Pull request by BarryNolte**
Better error handling when it comes to not having D2 CLI tools installed.

Fixes:
#33

### CODE REVIEW

**Review by gavin-ts — CHANGES_REQUESTED** [review]
haven't set up the CI for it yet but please run `yarn lint` and `yarn prettier -c .`

File: src/docToPreviewGenerator.ts [hunk 1]
```diff
@@ -149,4 +147,26 @@ export class DocToPreviewGenerator {
...
     return data;
   }
+
+  private strBreak =
+    "************************************************************";
+  private showErrorToolsNotFound(msg: string): void {
+    const errorMsgs: string[] = [
+      "D2 executable not found.",
+      "Make sure the D2 executable is installed and on system PATH.",
+      "https://d2lang.com",
```
Comment (gavin-ts) [inline src/docToPreviewGenerator.ts:157]:
```suggestion
      "https://d2lang.com/tour/install",
```

File: src/docToPreviewGenerator.ts [hunk 2]
```diff
@@ -149,4 +147,26 @@ export class DocToPreviewGenerator {
...
+      `${msg}`,
+    ];
+
+    outputChannel.appendError(this.strBreak);
+    for (const m in errorMsgs) {
+      if (errorMsgs[m]) {
+        outputChannel.appendError(errorMsgs[m]);
+      }
+    }
```
Comment (gavin-ts) [inline src/docToPreviewGenerator.ts:166]:
```suggestion
    for (const m of errorMsgs) {
      outputChannel.appendError(m);
    }
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of#difference_between_for...of_and_for...in

File: src/outputChannel.ts [hunk 3]
```diff
@@ -29,7 +29,18 @@ export class D2OutputChannel {
...
+
+    for (const m in msgs) {
+      if (msgs[m]) {
+        const output = msgs[m];
+        if (output.length === 0) {
+          continue;
+        }
+        this.outputChannel.appendLine(`[${time.join(":")}] - ${output}`);
+      }
+    }
```
Comment (gavin-ts) [inline src/outputChannel.ts:42]:
```suggestion
    for (const m of msgs) {
      if (m.length === 0) {
        continue;
      }
      this.outputChannel.appendLine(`[${time.join(":")}] - ${m}`);
    }
```

File: src/docToPreviewGenerator.ts [hunk 4]
```diff
@@ -149,4 +147,26 @@ export class DocToPreviewGenerator {
 
     return data;
   }
+
+  private strBreak =
+    "************************************************************";
+  private showErrorToolsNotFound(msg: string): void {
```
Comment (gavin-ts) [inline src/docToPreviewGenerator.ts:153]:
```suggestion
  private strBreak =
    "************************************************************";

  private showErrorToolsNotFound(msg: string): void {
```

lint:
```
d2-vscode/src/docToPreviewGenerator.ts
  153:3  error  Expected blank line between class members  lines-between-class-members

✖ 1 problem (1 error, 0 warnings)
```

**Review by gavin-ts — CHANGES_REQUESTED** [review]

File: src/outputChannel.ts (hunk 3 above)
Comment (BarryNolte) [reply src/outputChannel.ts]: I learn something new everyday... TypeScript is not my first language ;)

**Review by gavin-ts — APPROVED** [review]
can we also add the same check to the spawnSync in document formatter?

right now it just has this error in output after formatting without d2: `[16:00:17] - Cannot read properties of null (reading 'toString')`

**Comment by BarryNolte** [comment]
This brings up another option/opportunity.  There was discussion about making the preview generation async.  I think the 'vscode way' to do this is by making the generation a task, execute it asynchronously and fill in preview whenever it decides to be finished.  This gives the advantage that with tasks, you can have a 'problem matcher'  that automatically fills in the problem pane with line and column information so that it can be jumped to.  Playing around with it, it feels a little clunky, but does fit into the ecosystem better.  Thoughts?

**Comment by gavin-ts** [comment]
> This brings up another option/opportunity. There was discussion about making the preview generation async. I think the 'vscode way' to do this is by making the generation a task, execute it asynchronously and fill in preview whenever it decides to be finished. This gives the advantage that with tasks, you can have a 'problem matcher' that automatically fills in the problem pane with line and column information so that it can be jumped to. Playing around with it, it feels a little clunky, but does fit into the ecosystem better. Thoughts?

Yes I think adding a d2 compile task is the way to go. It could be used on its own manually, the preview could use it for its rendering, and even other extensions may be able to use the task. I can take a look if you have something you're trying out.

**Pull request merged**